### PR TITLE
output: Add backup feature for bad chunks

### PR DESCRIPTION
**Pull request by repeatedly**
Fix https://github.com/fluent/fluentd/issues/1856

Current implementation covers most bad chunk cases. I need feedback before writing test.

-  use `root_dir` parameter for backup directory when specified, `/${root_dir}/backup/worker${worker_id}/${plugin_id}/${chunk_id}.log`.
- Call secondary directly when unrecoverable error happens

Signed-off-by: Masahiro Nakagawa <repeatedly@gmail.com>

### CODE REVIEW

File: lib/fluent/plugin/output.rb [hunk 1]
```diff
@@ -1115,6 +1149,21 @@ def try_flush
         end
       end
 
+      def backup_chunk(chunk, using_secondary, delayed_commit)
+        unique_id = dump_unique_id_hex(chunk.unique_id)
+        safe_plugin_id = plugin_id.gsub(/[ \/\\:]/, '_')
```
Comment (cosmo0920) [inline lib/fluent/plugin/output.rb]: This feature assumes to work on Windows?
If so, we should add more prohibiting to use characters:

`/ > < ? : " \ * | ;` are not permitted to use on NTFS.

**Comment by cosmo0920** [comment]
I've added a comment about concerns on Windows environment.

**Comment by JohnMealy23** [comment]
The usage of this feature is hard to track down.  Can it be added to the docs?

**Comment by ganmacs** [comment]
@repeatedly Could you add the document about it. or had you already added?

**Comment by repeatedly** [comment]
Yes: https://docs.fluentd.org/buffer#handling-unrecoverable-errors

**Comment by jvshahid** [comment]
I Apologize for replying to an old issue. I am curious if there is a way to monitor the number of bad chunks, e.g., using Prometheus?

**Pull request merged**